ykrt/src/mt.rs

```diff
@@ -41,14 +45,62 @@ const PHASE_COUNTING: usize = 0b10; // The value specifies how many times we've
 /// mechanism is for every bytecode or AST node to have its own `Location` (even for bytecodes or
 /// nodes that can't be control points).
 #[derive(Debug)]
-pub struct Location {
-    pack: AtomicUsize,
+pub struct Location<I> {
+    // A Location is a state machine which operates as follows (where PHASE_COUNTING is the start
+    // state):
+    //
+    //               ┌────────────────────────────────────────┐
+    //               │                                        │ ─────────────┐
+    //   reprofile   │             PHASE_COUNTING             │              │
+    //  ┌──────────▶ │                                        │ ◀────────────┘
+    //  │            └────────────────────────────────────────┘    increment
+    //  │              │                     ▲         ▲           count
+    //  │              │ start tracing       │ abort   │ abort
+    //  │              ▼                     │         │
+    //  │            ┌────────────────────┐  │         │
+    //  │            │   PHASE_TRACING    │ ─┘         │
+    //  │            └────────────────────┘            │
+    //  │              │                               │
+    //  │              │ start compiling trace         │
+    //  │              │ in thread                     │
+    //  │              ▼                               │
+    //  │            ┌────────────────────┐            │
+    //  |            |                    | ───────────┘
+    //  │            │                    │
+    //  │            │  PHASE_COMPILING   │ ──────┐
+    //  │            │                    │       │ still compiling in thread
+    //  │            │                    │ ◀─────┘
+    //  │            └────────────────────┘
+    //  │              │
+    //  │              │ trace compiled
+    //  │              ▼
+    //  │            ┌────────────────────┐
+    //  │            │                    │ ──────┐
+    //  │            │   PHASE_COMPILED   │       │
+    //  └─────────── │                    │ ◀─────┘
+    //               └────────────────────┘
+    //
+    // We hope that a Location soon reaches PHASE_COMPILED (aka "the happy state") and stays there.
+    state: AtomicUsize,
+    phantom: PhantomData<I>,
```

Comment:
The phantom is purely for additional type safety? To stop us using locations for one control point with another?

Reply:
Yes and yes.